File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    TreeNode p, q;
    Stack<TreeNode> mem = new Stack<>(); // mem traveral
    Stack<TreeNode> pStack;
    Stack<TreeNode> qStack;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        this.p = p;
        this.q = q;
        traveral(root);
        while(pStack.size() > qStack.size()) {
            pStack.pop();
        }

        while(pStack.size() < qStack.size()) {
            qStack.pop();
        }

        while(!pStack.isEmpty() && !qStack.isEmpty()) {
            TreeNode pNode = pStack.pop();
            TreeNode qNode = qStack.pop();
            if(pNode.val == qNode.val) {
                return pNode;
            }
        }
        return null;
    }

    private void traveral(TreeNode node) {
        if(node == null) {
            return;
        }
        mem.push(node);
        if(p.val == node.val) {
            pStack = (Stack<TreeNode>)mem.clone();
        } else if(q.val == node.val) {
            qStack = (Stack<TreeNode>)mem.clone();
        }
        traveral(node.left);
        traveral(node.right);
        mem.pop();
    }
}

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
// ref : https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/solutions/5668568/video-solution-intuition-explained-in-detail-c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL || p == root || q == root) {
            return root;
        }
        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);
        if(left != NULL && right != NULL) { // case : p = 5, q = 1
            return root;
        } else if(left != NULL) {
            // not child p,q in right
            return left;
        } else { // right != NULL
            return right;
        }
    }
};